man/EquidistantTest.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "EQUIDISTANTTEST 1"
.TH EQUIDISTANTTEST 1 "2011-02-22" "GeographicLib 1.8" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
EquidistantTest \-\- perform projections based on geodesics
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBEquidistantTest\fR ( \fB\-z\fR | \fB\-c\fR | \fB\-g\fR ) \fIlat0\fR \fIlon0\fR [ \fB\-r\fR ]
[ \fB\-e\fR \fIa\fR \fIr\fR ] [ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Perform projections based on geodesics.  Convert geodetic coordinates to
either azimuthal equidistant, Cassini-Soldner, or gnomonic coordinates.
The center of the projection (\fIlat0\fR, \fIlon0\fR) is specified by either the \fB\-c\fR
option (for Cassini-Soldner), the \fB\-z\fR option (for azimuthal equidistant),
or the \fB\-g\fR option (for gnomonic).  At least one of these options must be
given (the last one given is used).
.PP
Geodetic coordinates are provided on standard input as a set of lines
containing (blank separated) \fIlatitude\fR and \fIlongitude\fR (decimal
degrees or degrees, minutes, seconds).  For each set of geodetic
coordinates, the corresponding projected coordinates \fIx\fR, \fIy\fR (meters)
are printed on standard output together with the azimuth \fIazi\fR
(degrees) and reciprocal scale \fIrk\fR.  For Cassini-Soldner, \fIazi\fR is
the bearing of the easting direction and the scale in the easting
direction is 1 and the scale in the northing direction is 1/\fIrk\fR.  For
azimuthal equidistant and gnomonic, \fIazi\fR is the bearing of the radial
direction and the scale in the azimuthal direction is 1/\fIrk\fR.  For
azimuthal equidistant and gnomonic, the scales in the radial direction
are 1 and 1/\fIrk\fR^2, respectively.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-z\fR" 4
.IX Item "-z"
use the azimuthal equidistant projection centered at latitude = \fIlat0\fR,
longitude = \fIlon0\fR.
.IP "\fB\-c\fR" 4
.IX Item "-c"
use the Cassini-Soldner projection centered at latitude = \fIlat0\fR,
longitude = \fIlon0\fR.
.IP "\fB\-g\fR" 4
.IX Item "-g"
use the ellipsoidal gnomonic projection centered at latitude = \fIlat0\fR,
longitude = \fIlon0\fR.
.IP "\fB\-r\fR" 4
.IX Item "-r"
perform the reverse projection.  \fIx\fR and \fIy\fR are given on standard
input and each line of standard output gives \fIlatitude\fR, \fIlongitude\fR,
\&\fIazi\fR, and \fIrk\fR.
.IP "\fB\-e\fR" 4
.IX Item "-e"
specify the ellipsoid via \fIa\fR \fIr\fR; the equatorial radius is \fIa\fR and
the reciprocal flattening is \fIr\fR.  Setting \fIr\fR = 0 results in a
sphere.  Specify \fIr\fR < 0 for a prolate ellipsoid.  By default, the
\&\s-1WGS84\s0 ellipsoid is used, \fIa\fR = 6378137m, \fIr\fR = 298.257223563.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
print version.
.IP "\fB\-h\fR" 4
.IX Item "-h"
print usage.
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
print full documentation.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Vb 4
\&   echo 48.648 \-2.007 | EquidistantTest \-c 48.836 2.337
\&   => \-319919 \-11791 86.7 0.999
\&   echo \-319919 \-11791 | EquidistantTest \-c 48.836 2.337 \-r
\&   => 48.648 \-2.007 86.7 0.999
.Ve
.SH "ERRORS"
.IX Header "ERRORS"
An illegal line of input will print an error message to standard output
beginning with \f(CW\*(C`ERROR:\*(C'\fR and causes \fBEquidistantTest\fR to return an exit
code of 1.  However, an error does not cause \fBEquidistantTest\fR to
terminate; following lines will be converted.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBEquidistantTest\fR is a part of GeographicLib,
<http://geographiclib.sf.net>.  The ellipsoidal gnomonic projection is
derived in Section 13 of C. F. F. Karney, \fIGeodesics on an ellipsoid of
revolution\fR, Feb. 2011; preprint <http://arxiv.org/abs/1102.1215>.
.SH "AUTHOR"
.IX Header "AUTHOR"
\&\fBEquidistantTest\fR was written by Charles Karney.


NEWS
# $Id$

A reverse chronological list of changes to GeographicLib

For more information, see

    http://geographiclib.sourceforge.net/

The current version of the library is 1.9.

Changes between 1.9 (released 2011-mm-dd) and 1.8 versions:

  * FIX BUG in area returned by Planimeter for pole encircling polygons.

  * FIX BUG in error message reported when DMS::Decode reads the string
    "5d.".

  * FIX BUG in AlbersEqualArea::Reverse (lon0 not being used).

  * Ensure that all exceptions thrown in the Utility Programs are
    caught.

  * Avoid using catch within GeographicLib::DMS.

  * Move Accumulator class from Planimeter.cpp to Constants.hpp.

  * Add Math::sq<T>.

  * Simplify Installing the geoid datasets
    + add geographiclib-get-geoids for Unix-like systems;
    + add installers for Windows.

  * Provide cmake support:
    + build binary installer for Windows;
    + include regression tests;
    + add --input-string, --input-file, --output-file options to the
      Utility Programs to support tests.

  * Rename utility EquidistantTest as GeodesicProj and
    TransverseMercatorTest as TransverseMercatorProj.

  * Add ConicProj.

  * Reverse the initial sense of the -s option for Planimeter.

  * Migrate source from subversion to git.

Changes between 1.8 (released 2011-02-22) and 1.7 versions:

  * Optionally return rotation matrix from GeographicLib::Geocentric and
    GeographicLib::LocalCartesian.

  * For the Utility Programs, supply man pages, -h prints the synopsis,
    --help prints the man page, --version prints the version.

  * Use accurate summation in Planimeter.

  * Add 64-bit targets for Visual Studio 2010.

  * Use templates for defining math functions and some constants.

  * GeographicLib::Geoid updates
    + Add GeographicLib::Geoid::DefaultGeoidPath and
      GeographicLib::Geoid::DefaultGeoidName;
    + GeoidEval uses environment variable GEOID_NAME as the default
      geoid;
    + Add --msltohae and --haetomsl as GeoidEval options (and don't
      document the single hyphen versions).

  * Remove documentation that duplicates papers on transverse Mercator
    and geodesics.

Changes between 1.7 (released 2010-12-21) and 1.6 versions:

  * FIX BUG in scale returned by GeographicLib::LambertConformalConic::Reverse.

  * Add GeographicLib::AlbersEqualArea projection.

  * Library created by Visual Studio is Geographic.lib instead of
    GeographicLib.lib (compatible with makefiles).

  * Make classes NaN aware.

  * Use cell arrays for MGRS strings in Matlab.

  * Add solution/project files for Visual Studio 2010 (32-bit only).

  * Use C++0X static_assert and math functions, if available.

Change between 1.6 (released 2010-11-23) and 1.5 versions:

  * FIX BUG introduced in GeographicLib::Geoid in version 1.5 (found by
    Dave Edwards).

Changes between 1.5 (released 2010-11-19) and 1.4 versions:

  * Improve area calculations for small polygons.

  * Add -s and -r flags to Planimeter utility.

  * Improve the accuracy of GeographicLib::LambertConformalConic using
    divided differences.

  * FIX BUG in meridian convergence returned by
    LambertConformalConic::Forward.

  * Add optional threadsafe parameter to GeographicLib::Geoid
    constructor.  WARNING: This changes may break binary compatibility
    with previous versions of GeographicLib.  However, the library is
    source compatible.

  * Add GeographicLib::OSGB.

  * Matlab and Octave interfaces to GeographicLib::UTMUPS,
    GeographicLib::MGRS, GeographicLib::Geoid, GeographicLib::Geodesic
    provided.

  * Minor changes
    + explicitly turn on optimization in Visual Studio 2008 projects;
    + add missing dependencies in some Makefiles;
    + move pi() and degree() from GeographicLib::Constants to
      GeographicLib::Math;
    + introduce GeographicLib::Math::extended type to aid testing;
    + add GeographicLib::Math::epi() and GeographicLib::Math::edegree().
    + fixes to compile under cygwin;
    + tweak expression used to find latitude from conformal latitude.

Changes between 1.4 (released 2010-09-12) and 1.3 versions:

  * Changes to GeographicLib::Geodesic and GeographicLib::GeodesicLine:
    + FIX BUG in Geodesic::Inverse with prolate ellipsoids;
    + add area computations to Geodesic::Direct and Geodesic::Inverse;
    + add geodesic areas to geodesic test set;
    + make GeodesicLine constructor public;
    + change longitude series in Geodesic into Helmert-like form;
    + ensure that equatorial geodesics have cos(alpha0) = 0 identically;
    + generalize interface for Geodesic and GeodesicLine;
    + split GeodesicLine and Geodesic into different files;
    + signal convergence failure in Geodesic::Inverse with NaNs;
    + deprecate one function in Geodesic and two functions in
      GeodesicLine;
    + deprecate -n option for Geod.
    WARNING: These changes may break binary compatibility with previous
    versions of GeographicLib.  However, the library is source
    compatible (with the proviso that GeographicLib/GeodesicLine.hpp may
    now need to be included).

  * Add the Planimeter utility for computing the areas of
    geodesic polygons.

  * Improve iterative solution of GeographicLib::Gnomonic::Reverse.

  * Add GeographicLib::Geoid::ConvertHeight.

  * Add -msltohae, -haetomsl, and -z options to \ref geoideval.

  * Constructors check that minor radius is positive.

  * Add overloaded Forward and Reverse functions to the projection
    classes which don't return the convergence (or azimuth) and scale.

  * Document function parameters and return values consistently.

Changes between 1.3 (released 2010-07-21) and 1.2 versions:

  * Add GeographicLib::Gnomonic, the ellipsoid generalization of the
    gnomonic projection.

  * Add -g and -e options to Equidistanttest.

  * Use fixed-point notation for output from Cartconvert,
    Equidistanttest, Transversemercatortest.

  * PolarStereographic:
    + Improved conversion to conformal coordinates;

tools/EquidistantTest.cpp
/**
 * \file EquidistantTest.cpp
 * \brief Command line utility for azimuthal equidistant and Cassini-Soldner
 * projections
 *
 * Copyright (c) Charles Karney (2009, 2010, 2011) <charles@karney.com>
 * and licensed under the LGPL.  For more information, see
 * http://geographiclib.sourceforge.net/
 *
 * Compile with -I../include and link with Geodesic.o GeodesicLine.o
 * AzimuthalEquidistant.o Gnomonic.o CassiniSoldner.o
 *
 * See the <a href="EquidistantTest.1.html">man page</a> for usage
 * information.
 **********************************************************************/

#include "GeographicLib/Geodesic.hpp"
#include "GeographicLib/AzimuthalEquidistant.hpp"
#include "GeographicLib/CassiniSoldner.hpp"
#include "GeographicLib/Gnomonic.hpp"
#include "GeographicLib/DMS.hpp"
#include <iostream>
#include <sstream>

#include "EquidistantTest.usage"

int main(int argc, char* argv[]) {
  using namespace GeographicLib;
  typedef Math::real real;
  bool azimuthal = false, cassini = false, gnomonic = false, reverse = false;
  real lat0 = 0, lon0 = 0;
  real
    a = Constants::WGS84_a<real>(),
    r = Constants::WGS84_r<real>();
  for (int m = 1; m < argc; ++m) {
    std::string arg(argv[m]);
    if (arg == "-r")
      reverse = true;
    else if (arg == "-c" || arg == "-z" || arg == "-g") {
      cassini = arg == "-c";
      azimuthal = arg == "-z";
      gnomonic = arg == "-g";
      if (m + 2 >= argc) return usage(1, true);
      try {
        DMS::DecodeLatLon(std::string(argv[m + 1]), std::string(argv[m + 2]),
                          lat0, lon0);
      }
      catch (const std::exception& e) {
        std::cerr << "Error decoding arguments of " << arg << ": "
                  << e.what() << "\n";
        return 1;
      }
      m += 2;
    } else if (arg == "-e") {
      if (m + 2 >= argc) return usage(1, true);
      try {
        a = DMS::Decode(std::string(argv[m + 1]));
        r = DMS::Decode(std::string(argv[m + 2]));
      }
      catch (const std::exception& e) {
        std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
        return 1;
      }
      m += 2;
    } else if (arg == "--version") {
      std::cout
        << PROGRAM_NAME
        << ": $Id: EquidistantTest.cpp 6978 2011-02-21 22:42:11Z karney $\n"
        << "GeographicLib version " << GEOGRAPHICLIB_VERSION << "\n";
      return 0;
    } else
      return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
  }

  if (!(azimuthal || cassini || gnomonic)) {
    std::cerr << "Must specify \"-z lat0 lon0\" or "
              << "\"-c lat0 lon0\" or \"-g lat0 lon0\"\n";
    return 1;
  }

  const Geodesic geod(a, r);
  const CassiniSoldner cs = cassini ?
    CassiniSoldner(lat0, lon0, geod) : CassiniSoldner(geod);
  const AzimuthalEquidistant az(geod);
  const Gnomonic gn(geod);

  std::string s;
  int retval = 0;
  std::cout << std::fixed;
  while (std::getline(std::cin, s)) {
    try {
      std::istringstream str(s);
      real lat, lon, x, y, azi, rk;
      std::string stra, strb;
      if (!(str >> stra >> strb))
        throw GeographicErr("Incomplete input: " + s);
      if (reverse) {
        x = DMS::Decode(stra);
        y = DMS::Decode(strb);
      } else
        DMS::DecodeLatLon(stra, strb, lat, lon);
      std::string strc;
      if (str >> strc)
        throw GeographicErr("Extraneous input: " + strc);
      if (reverse) {
        if (cassini)
          cs.Reverse(x, y, lat, lon, azi, rk);
        else if (azimuthal)
          az.Reverse(lat0, lon0, x, y, lat, lon, azi, rk);
        else
          gn.Reverse(lat0, lon0, x, y, lat, lon, azi, rk);
        std::cout << DMS::Encode(lat, 15, DMS::NUMBER) << " "
                  << DMS::Encode(lon, 15, DMS::NUMBER) << " "
                  << DMS::Encode(azi, 15, DMS::NUMBER) << " "
                  << DMS::Encode(rk, 16, DMS::NUMBER) << "\n";
      } else {
        if (cassini)
          cs.Forward(lat, lon, x, y, azi, rk);
        else if (azimuthal)
          az.Forward(lat0, lon0, lat, lon, x, y, azi, rk);
        else
          gn.Forward(lat0, lon0, lat, lon, x, y, azi, rk);
        std::cout << DMS::Encode(x, 10, DMS::NUMBER) << " "
                  << DMS::Encode(y, 10, DMS::NUMBER) << " "
                  << DMS::Encode(azi, 15, DMS::NUMBER) << " "
                  << DMS::Encode(rk, 16, DMS::NUMBER) << "\n";
      }
    }
    catch (const std::exception& e) {
      std::cout << "ERROR: " << e.what() << "\n";
      retval = 1;
    }
  }

  return retval;
}
